liblwgeom/lwgeom_geos.c

```diff
@@ -495,224 +490,180 @@ lwgeom_geos_version()
+	if (geom1->srid > SRID_MAXIMUM)
 	{
-		lwerror("Error in GEOSNormalize: %s", lwgeom_geos_errmsg);
-		return NULL; /* never get here */
+		lwerror("%s: SRID is more than maximum (%u > %u)", funcname, geom1->srid, SRID_USER_MAXIMUM);
+		return SRID_INVALID;
 	}
+	return geom1->srid;
+}
 
-	GEOSSetSRID(g1, srid); /* needed ? */
-	result = GEOS2LWGEOM(g1, is3d);
-	GEOSGeom_destroy(g1);
-
-	if (!result)
+/* Input decoder and sanity checker for GEOS wrappers */
+inline static const uint32_t
+input_lwgeom_to_geos(GEOSGeometry** g, const LWGEOM* geom, const char* funcname)
+{
+	*g = LWGEOM2GEOS(geom, 1);
+	if (!*g)
 	{
-		lwerror("Error performing normalize: GEOS2LWGEOM: %s", lwgeom_geos_errmsg);
-		return NULL; /* never get here */
+		lwerror("%s: argument geometry could not be converted to GEOS: %s", funcname, lwgeom_geos_errmsg);
+		return LW_FALSE;
 	}
-	return result;
+	return LW_TRUE;
 }
 
-LWGEOM*
-lwgeom_intersection(const LWGEOM* geom1, const LWGEOM* geom2)
+/* Output encoder and sanity checker for GEOS wrappers */
+inline static const uint32_t
+output_geos_as_lwgeom(GEOSGeometry** g, LWGEOM** geom, const uint32_t srid, const uint32_t is3d, const char* funcname)
 {
-	LWGEOM* result;
-	GEOSGeometry *g1, *g2, *g3;
-	int is3d;
-	int srid;
-
-	/* A.Intersection(Empty) == Empty */
-	if (lwgeom_is_empty(geom2)) return lwgeom_clone_deep(geom2);
+	GEOSSetSRID(*g, srid);
+	*geom = GEOS2LWGEOM(*g, is3d);
+	if (!*geom)
+	{
+		lwerror("%s: result geometry could not be converted from GEOS: %s", funcname, lwgeom_geos_errmsg);
+		return LW_FALSE;
+	}
+	return LW_TRUE;
+}
 
-	/* Empty.Intersection(A) == Empty */
-	if (lwgeom_is_empty(geom1)) return lwgeom_clone_deep(geom1);
+/* Inintialize GEOS and first GEOSGeometry in wrapper in one line */
+inline static GEOSGeometry*
+init_geos_geometry_and_return_null()
+{
+	initGEOS(lwnotice, lwgeom_geos_error);
+	return NULL;
+}
 
-	/* ensure srids are identical */
-	srid = (int)(geom1->srid);
-	error_if_srid_mismatch(srid, (int)(geom2->srid));
+/* Output encoder and sanity checker for GEOS wrappers */
+inline static LWGEOM*
+geos_clean_and_fail(GEOSGeometry** g1, GEOSGeometry** g2, GEOSGeometry** g3, const char* funcname)
+{
+	if (g1) GEOSGeom_destroy(*g1);
+	if (g2) GEOSGeom_destroy(*g2);
+	if (g3) GEOSGeom_destroy(*g3);
+	lwerror("%s: GEOS Error: %s", funcname, lwgeom_geos_errmsg);
+	return NULL;
+}
 
-	is3d = (FLAGS_GET_Z(geom1->flags) || FLAGS_GET_Z(geom2->flags));
+/* Output encoder and sanity checker for GEOS wrappers */
+inline static void
+geos_clean(GEOSGeometry** g1, GEOSGeometry** g2, GEOSGeometry** g3)
+{
+	if (g1) GEOSGeom_destroy(*g1);
+	if (g2) GEOSGeom_destroy(*g2);
+	if (g3) GEOSGeom_destroy(*g3);
+	return;
+}
 
-	initGEOS(lwnotice, lwgeom_geos_error);
+LWGEOM*
+lwgeom_normalize(const LWGEOM* geom)
+{
+	LWGEOM* result;
+	uint32_t srid = get_result_srid(geom, NULL, __func__);
+	uint32_t is3d = FLAGS_GET_Z(geom->flags);
+	GEOSGeometry* g = init_geos_geometry_and_return_null();
```

Comment:
This seems like a confusing way to go from two lines to one.

Reply:
It's true. It looked good on first two functions though :)
Changed it to call `initGEOS` explicitly.